### PR TITLE
Dynamic Cluster Pool

### PR DESCRIPTION
**Pull request by gandro**
This adds a new IPAM mode, named `cluster-pool-v2`, where the operator is able to assign and remove additional pod CIDRs to a node based on demand. The current usage of pod CIDRs is indicated by the Cilium Agent in the local CiliumNode CRD:

```
This extends the CiliumNode CRD with two spec and one status fields
needed to support dynamic allocation of pod CIDRs to node.

  - `CiliumNode.Status.IPAM.UsedPodCIDRs` informs the operator about the
    pod CIDR usage of this node.  Each pod CIDR can be in one of three
    states:

    * `in-use`: This pod CIDR is currently being used for allocation of
      pod IPs.
    * `depleted`: Every IP in this pod CIDR is currently used up. If all
      pod CIDRs are depleted, the operator will allocate and assign new
      pod CIDRs to this node.
    * `released`: This pod CIDR is no longer in use by the agent and
      should be deallocated by the operator.

  - `CiliumNode.Spec.IPAM.PodCIDRAllocationThreshold` defines the
    minimum number of free IPs which must be available to this node via
    its pod CIDR pool. If the total number of IP addresses in the pod CIDR
    pool is less than this value, the pod CIDRs currently in-use by this
    node will be marked as depleted and cilium-operator will allocate a new
    pod CIDR to this node.  This value effectively defines the buffer of IP
    addresses available immediately without requiring cilium-operator to get
    involved.

  - `CiliumNode.Spec.IPAM.PodCIDRReleaseThreshold` defines the maximum
    number of free IPs which may be available to this node via its pod
    CIDR pool. While the total number of free IP addresses in the pod CIDR
    pool is larger than this value, cilium-agent will attempt to release
    currently unused pod CIDRs.
```

See individual commit messages for more details.

This version of the feature is in beta state. A follow-up PR will add
- Documentation
- CI test(s) to complement the existing unit test

*Please indicate in the review what feedback is a blocker for rc0, and which feedback may be addressed in a follow-up PR.*

```release-note
Support dynamic allocation of pod CIDRs in cluster pool v2 IPAM mode
```

### CODE REVIEW

File: pkg/ipam/clusterpool.go
Comment (twpayne) [inline pkg/ipam/clusterpool.go]: Do we still need the special case handling of the first pod CIDR? IIRC at one point we decided it was unnecessary.

File: pkg/ipam/clusterpool.go
Comment (twpayne) [inline pkg/ipam/clusterpool.go]: `pkg/k8s.IsLocalCiliumNode` could be used instead of this function.

File: pkg/ipam/allocator/podcidr/podcidr_v2.go
Comment (twpayne) [inline pkg/ipam/allocator/podcidr/podcidr_v2.go]: Consider using `go.uber.org/multierr` (already indirectly imported) instead.

File: pkg/ipam/clusterpool.go
Comment (gandro) [reply pkg/ipam/clusterpool.go]: Yes, you are absolutely right. You reviewed it a second too soon. I have already removed it locally, but haven't pushed yet because I was waiting on CI to finish. Will push an updated version soon.

File: pkg/ipam/allocator/podcidr/podcidr_v2.go
Comment (gandro) [reply pkg/ipam/allocator/podcidr/podcidr_v2.go]: Great suggestion, looks much cleaner now!

File: pkg/datapath/linux/node.go
Comment (christarazi) [inline pkg/datapath/linux/node.go:253]: I reviewed the logic in this function and it LGTM in terms of logical gaps from the old implementation. But to be really sure, I noticed that we are lacking some unit tests for this specific function so it might be worth writing unit tests _against the old_ version first, then apply this commit and see if anything breaks (of course there will be slight modifications required as the func params changed, etc).

**Review by christarazi — COMMENTED** [review]
As discussed offline, the overall PR looks good as a first go. My main focus was to look at the first couple of commits to identify any changes that may have caused regressions as the rest of the changes are guarded by the fact that the user must opt into running the cluster-pool-v2 mode, so the risk is mitigated. A look at the latter commits especially the last couple is deferred for now. 

With that said, I did leave a comment re: the second commit.

File: pkg/datapath/linux/node.go
Comment (gandro) [reply pkg/datapath/linux/node.go:253]: Thanks! That's a really good suggestion. Will implement asap.

Comment (gandro) [reply pkg/datapath/linux/node.go:253]: Turns out, we already have a pretty good unit test for the function: https://github.com/cilium/cilium/blob/6237e74b4f0ffbfc38b5c26f5a0a8f94f33a6c69/pkg/datapath/linux/node_linux_test.go#L646

Doesn't seem like it regressed (run-time tests passed before). But I'll extend it to make sure we cover all corner cases and add support the new behavior.

**Review by twpayne — APPROVED** [review]

**Comment by gandro** [comment]
Last push added some more unit tests for the direct routing code and some comments on the agent's side of IPAM. [See diff](https://github.com/cilium/cilium/compare/7f47aa25347c807f9daa6ca3279b0aaba51134a4..d646e1a17356d08f14e57e48da10646187510f9f)

File: pkg/datapath/linux/node.go
Comment (gandro) [reply pkg/datapath/linux/node.go:253]: Extended the unit tests accordingly: https://github.com/cilium/cilium/pull/18887/commits/d646e1a17356d08f14e57e48da10646187510f9f

**Review by christarazi — APPROVED** [review]
:rocket:

**Comment by michi-covalent** [comment]
/test-1.16-4.9

File: pkg/k8s/watchers/cilium_node.go
Comment (aanm) [inline pkg/k8s/watchers/cilium_node.go:71]: Why aren't we comparing the local node equalness to avoid consuming unnecessary CPU cycles?

File: pkg/defaults/defaults.go
Comment (aanm) [inline pkg/defaults/defaults.go:349]: Might be worth just adding a small comment here on why 8 was picked.

File: pkg/defaults/defaults.go
Comment (aanm) [inline pkg/defaults/defaults.go:353]: Might be worth just adding a small comment here on why 16 was picked.

File: pkg/ipam/types/types.go
Comment (aanm) [inline pkg/ipam/types/types.go:140]: The `used-pod-cidr` could be renamed to `pod-cidr` because we have a field status for each pod-cidr. Having `used` in here might be misleading if a pod cidr was "`released`: This pod CIDR is no longer in use by the agent and should be deallocated by the operator.", which begs the question is this podCIDR "used" or "released"?

File: pkg/ipam/option/option.go
Comment (aanm) [inline pkg/ipam/option/option.go:27]: `cluster-pool-v2` sounds good for now, we can deprecate the old behavior of `cluster-pool` if `cluster-pool-v2` is solid for the v1.12 release

File: pkg/ipam/clusterpool.go
Comment (aanm) [inline pkg/ipam/clusterpool.go:454]: Not sure, maybe it is worth double check that `allocationThreshold` is never less than `releaseThreshold`?

**Review by aanm — APPROVED** [review]
Awesome work :tada: Only some comments, none of them are blocking and can be addressed afterwards.

File: pkg/node/address.go
Comment (joestringer) [inline pkg/node/address.go:713]: I know that this style is more consistent with the way that most of `pkg/node` works, but I was hoping not to expand on this model. We used this style quite a bit across the Cilium codebase, and it's often one of the reasons that unit testing is difficult, because you can't just declare an interface & mock/dependency-inject such code.

Looking at the caller, previously the dependency here was explicitly passed into the initialization function, but now it calls this getter and the ordering is less explicit around how this value gets populated.  I think it's harder to trace both (1) initialization + (2) updates to see we are correctly handling the updates, especially now that this is more dynamic.

Have you thought about this too? I don't think this is appropriate to have in the discussion around this particular PR, so I don't want to block this. I would like to start a thread of discussion on some code maintenance / code debt items that I see like this though. At least hopefully get on the same page on whether we think it continues to make sense to accept code in this style or not.

**Review by joestringer — APPROVED** [review]
I mostly just looked at health since the "datapath" (userspace node handling) bits were already reviewed by Chris. Just one non-blocking comment.

**Comment by michi-covalent** [comment]
this one looks like a legitimate failure: https://app.travis-ci.com/github/cilium/cilium/jobs/561487847#L15710 :mag: 
```
FAIL: podcidr_v2_test.go:20: PodCIDRSuite.TestNodesPodCIDRManager_allocateNodeV2
```

File: pkg/ipam/option/option.go
Comment (gandro) [reply pkg/ipam/option/option.go:27]: Yes, this is my thinking as well. There are still some known shortcomings with v2, such as the lack of IPSec support that need to be addressed first.

File: pkg/node/address.go
Comment (gandro) [reply pkg/node/address.go:713]: Yes, this also came up during the review session with Chris, Michi and Tom. I fully agree that we should get rid of the `pkg/node` pattern. Besides the problems you mentioned, we also we have two sources of truth for the local node (`pkg/node` and `pkg/nodediscovery`) which I think is a potential footgun.

My thinking here is also that we should get rid of the `pkg/node` getters. But I didn't want to tackle that refactor in this PR (both due to time constraints and the size of the PR). The sole reason I added getters to `pkg/node` instead of `pkg/nodediscovery.NodeDiscovery` is that at its more consistent with the existing code.

Let's create an issue or thread discussing how to address this technical debt in a follow-up PR.

**Comment by gandro** [comment]
> this one looks like a legitimate failure: https://app.travis-ci.com/github/cilium/cilium/jobs/561487847#L15710 mag
> 
> ```
> FAIL: podcidr_v2_test.go:20: PodCIDRSuite.TestNodesPodCIDRManager_allocateNodeV2
> ```

Interesting. It didn't fail before and doesn't locally. Might be indication that the test is flaky. Investigating.

File: pkg/k8s/watchers/cilium_node.go
Comment (gandro) [reply pkg/k8s/watchers/cilium_node.go:71]: You mean checking for full equality using `reflect.DeepEqual`? I can add that, depending on the subscribers it might not save us any cycles though.

**Comment by gandro** [comment]
/test-1.16-4.9

**Comment by gandro** [comment]
Seems like we hit https://github.com/cilium/cilium/issues/18977 again. That is suspicious. I'll do a triage to ensure this really is a flake, and not a regression introduced by this PR.

**Comment by gandro** [comment]
Did a triage of https://github.com/cilium/cilium/issues/18977 - my findings are inconclusive :confused: 

- I haven't been able to establish a root cause for the test failure.
- The failing test case uses tunnel-mode, so the more sensitive direct routing mode changes here are not to blame. My PR also changes the local node routes (which is used in tunneling mode), but those routes look okay in the sysdump.
- Looking at the CI dashboard, this seems to be a new failure which only has two occurrences in this PR and in PR#17758 - but in PR#17758 it failed due to previous test failures, so effectively this PR is the only PR where the test is failing.

image

Given that I don't have established a root cause for the failure, I'll open a canary PR to run the same test suite a few times against master. If the canary PR remains green, then this PR introduces a regression. If it fails with the same failure, then we have a new flake, but unrelated to this PR.

**Comment by gandro** [comment]
Canary PR doesn't seem to show the same issue. I'm still trying to find a root cause for the failure. Meanwhile I'll also run the test suite again.

**Comment by gandro** [comment]
/test-1.16-4.9

**Comment by michi-covalent** [comment]
/test-1.16-4.9

**Comment by gandro** [comment]
Did some more digging into the failure together with Martynas. While we still don't understand the root cause, there are reasons to assume that the failure is _not_ related to this PR:

- The test fails because a curl from host ("k8s1") to ClusterIP (10.100.84.120:10080) curl times out (exit code 28). My concern was that this might have been caused by my changes to the local node route code, as it's the only relevant change in this PR (the test case doesn't use direct routing or the new IPAM mode).
 However, the sysdump shows that the local node routes are set up correctly on both nodes (see below), and more importantly, we see successful connectivity before the failed one (in fact its failing in the last round, 10/10 as indicated by `failed: :4367/10=28`). Since in the test the local node route is never updated after initialization (and I see the required `msg="Received node update event from local` which adds the route a couple minutes before the test runs), this makes me confident that my local node route changes are not likely to be the reason for the connectivity failure.
- While the flake was first observed in this PR under its new name (`Checks E/W loadbalancing`) #18977, there is a very similar older CI flake https://github.com/cilium/cilium/issues/18530 under its old name.

Correct local node routes:
```console
$ grep cilium_host bugtool-cilium-*/cmd/ip--4-r.md  
bugtool-cilium-jzw8r/cmd/ip--4-r.md:10.0.1.0/24 via 10.0.1.142 dev cilium_host src 10.0.1.142 
bugtool-cilium-jzw8r/cmd/ip--4-r.md:10.0.0.0/24 via 10.0.1.142 dev cilium_host src 10.0.1.142 mtu 1450 
bugtool-cilium-jzw8r/cmd/ip--4-r.md:10.0.1.142 dev cilium_host scope link 

bugtool-cilium-kr5ph/cmd/ip--4-r.md:10.0.0.0/24 via 10.0.0.198 dev cilium_host src 10.0.0.198 
bugtool-cilium-kr5ph/cmd/ip--4-r.md:10.0.1.0/24 via 10.0.0.198 dev cilium_host src 10.0.0.198 mtu 1450 
bugtool-cilium-kr5ph/cmd/ip--4-r.md:10.0.0.198 dev cilium_host scope link 
```

**Review by rolinh — APPROVED** [review]
Due to the sheer size of the PR, I limited my review to the topic I was pulled in for: check vendor changes. Turned the review from a lengthy one to a 2s one :grin: 

PS: I like seeing so many tests :heart_eyes:

**Comment by gandro** [comment]
Marking as ready-to-merge to unblock rc0. 

- I will address the remaining feedback in a follow-up PR. Thanks y'all for the review!
- I will keep an eye on #18977 in the CI dashboard

**Comment by joestringer** [comment]
I just noticed this was marked `release-note/minor`, but we may want to bump it to `release-note/major` for the final v1.12 release as I think this is one of those bigger-ticket items that we would typically blog about & try to get people interested in trying it out.

File: pkg/ipam/allocator/podcidr/podcidr.go
Comment (kaworu) [inline pkg/ipam/allocator/podcidr/podcidr.go:385]: typo?
```
 		// ClusterPool v1 never returns both the spec and the status
```

File: pkg/cidr/cidr_test.go
Comment (kaworu) [inline pkg/cidr/cidr_test.go:298]: nitpicking: as I understand it `RemoveAll` modify its first slice parameter by design, and if correct then I think this behaviour should be tested as well.

>  		c.Assert(tt.args.ipNets, checker.DeepEquals, tt.want, check.Commentf("Test Name: %s", tt.name))

File: pkg/ipam/allocator/podcidr/podcidr.go
Comment (gandro) [reply pkg/ipam/allocator/podcidr/podcidr.go:385]: Yes, thanks!

File: pkg/cidr/cidr_test.go
Comment (gandro) [reply pkg/cidr/cidr_test.go:298]: Yes, `RemoveAll` clobbers its first argument (good point, I should mention that in the doc comment) and must not be used after calling the function - but I don't want to guarantee that the input argument equals the result (similar to e.g. how `append` works) - a future implementation might optimize memory consumption by re-allocating and shrinking the capacity of the resulting slice. In that case, the return value will be different from the input argument.

**Comment by qmonnet** [comment]
For the record: Some commits of this PR are being backported to 1.10, 1.11 in the PRs linked above.

**Pull request merged**
